WIP: Starknet Snap Rough Skeleton (#1509)
MetaMask/metamask-docs: wallet/how-to/use-non-evm-networks/starknet/manage-starknet-networks.md

---
description: Manage Starknet networks in MetaMask.
sidebar_position: 3
---

import Tabs from "@theme/Tabs";
import TabItem from "@theme/TabItem";

# Manage Starknet networks

You can detect a user's Starknet network and prompt them to switch Starknet networks in MetaMask,
using the
[`get-starknet`](https://github.com/starknet-io/get-starknet) library or the
[`wallet_invokeSnap`](/snaps/reference/wallet-api-for-snaps/#wallet_invokesnap) JSON-RPC method.

## Prerequisites

[Connect to Starknet](connect-to-starknet.md) from your dapp.

## Detect a user's network

Detect the Starknet network a user is currently connected to using the following:

<Tabs>
  <TabItem value="get-starknet" default>

  ```javascript

>   const checkCurrentNetwork = () => {
>     const starknet = getStarknet();
>     const currentNetwork = starknet.provider.rpcUrl;
>     console.log("Currently connected to:", currentNetwork);
>     return currentNetwork;
>   };

Reviewer: ```suggestion
  const checkCurrentNetwork = (wallet) => {
      try {
        if(wallet?.isConnected !== true){
          throw("Wallet not connected");
        } 
        const currentNetwork = wallet?.chchainId
        console.log("Currently connected to:", currentNetwork);
        return currentNetwork;
      } catch (error) {
        console.error("Error of detect current connected network:", error);
      }
  };
```

we should return the chainId , rather than rpcUrl, and standardized the example method as other

  ```
  
  </TabItem>
  <TabItem value="wallet_invokeSnap">

  ```javascript

>   const checkCurrentNetwork = async () => {
>     if (typeof window.ethereum !== "undefined" && window.ethereum.isMetaMask) {
>       try {
>         const response = await window.ethereum.request({
>           method: "wallet_invokeSnap",
>           params: {
>             snapId: "npm:@consensys/starknet-snap",
>             request: {
>               method: "starknet_getChainId"
>             }
>           }
>         });
>         
>         let networkName;
>         switch (response) {
>           case "0x534e5f4d41494e":
>             networkName = "Mainnet";
>             break;
>           case "0x534e5f5345504f4c4941":
>             networkName = "Sepolia Testnet";
>             break;
>           default:
>             networkName = "Unknown Network";
>         }
>         
>         console.log("Currently connected to:", networkName);
>         return response;  // Returns the chain ID.
>       } catch (error) {
>         console.error("Error getting current Starknet network:", error);
>         throw error;

Reviewer: ```suggestion
  const checkCurrentNetwork = async () => {
    const provider = await getEip6963Provider()

    if (provider) {
      try {
        const response = await provider.request({
          method: "wallet_invokeSnap",
          params: {
            snapId: "npm:@consensys/starknet-snap",
            request: {
              method: "starkNet_getCurrentNetwork"
            }
          }
        });
        
        console.log("Currently connected to:", response.name);
        return response.chainId;  // Returns the chain ID.
      } catch (error) {
        console.error("Error getting current Starknet network:", error);
        throw error;
```

we should always guide user to use eip6963 to  get the provider, instead of using window.ethereum
hence i create a mock implementation `const provider = await getEip6963Provider()`

Author: This is a bit confusing @stanleyyconsensys. We can add a note to use the recommended EIP-6963, but for all the other methods we are using `window.ethereum`. For consistency could we provide both, or maybe I can include a note to flag this,

      }
    } else {
      console.error("MetaMask not detected or Snaps not supported");
      throw new Error("MetaMask not detected or Snaps not supported");
    }
  };

  checkCurrentNetwork().then(chainId => {
    console.log("Chain ID:", chainId);
  }).catch(error => {
    console.error("Error:", error);
  });
  ```

  </TabItem> 
</Tabs>
  
## Switch networks

Starknet currently supports two public networks, Mainnet and Sepolia testnet.
Prompt users to switch between networks by setting the
[chain ID](../../../reference/non-evm-apis/starknet-snap-api.md#supported-networks) of the target network:

<Tabs>
  <TabItem value="get-starknet" default>

  ```javascript

>   const switchChain = async (chainId: string) => {
>     try {
>       await wallet?.request({
>         type: "wallet_switchStarknetChain",
>         params: { chainId: chainId },
>       });
>         console.log(`Switched to chainId: ${chainId}`);
>     } catch (e) {
>       console.error("Failed to switch chain:", e);
>     }
>   };

Reviewer: ```suggestion
  const switchChain = async (wallet, chainId) => {
    try {
      if(wallet?.isConnected !== true){
          throw("Wallet not connected");
      }
       
      await wallet?.request({
        type: "wallet_switchStarknetChain",
        params: { chainId: chainId },
      });
      console.log(`Switched to chainId: ${chainId}`);
    } catch (e) {
      console.error("Failed to switch chain:", e);
    }
  };
```

standardised with other example with an wallet object and check if it has connected before execute (yes, switch network require connected as well)

  ```

  </TabItem>
  <TabItem value="wallet_invokeSnap">

  ```javascript

>   const switchStarknetNetwork = async (chainId) => {
>     if (typeof window.ethereum !== "undefined" && window.ethereum.isMetaMask) {
>       try {
>         await window.ethereum.request({
>           method: "wallet_invokeSnap",
>           params: {
>             snapId: "npm:@consensys/starknet-snap",
>             request: {
>               method: "wallet_switchStarknetChain",
>               params: { chainId: chainId }
>             }
>           }
>         });
>         console.log(`Switched to Starknet network with chainId: ${chainId}`);
>       } catch (error) {
>         console.error("Error switching Starknet network:", error);
>         throw error;
>       }
>     } else {
>       console.error("MetaMask not detected or Snaps not supported");
>       throw new Error("MetaMask not detected or Snaps not supported");
>     }
>   };

Reviewer: ```suggestion
  const switchStarknetNetwork = async (chainId) => {
    const provider = await getEip6963Provider()

    if (provider) {
      try {
        await provider.request({
          method: "wallet_invokeSnap",
          params: {
            snapId: "npm:@consensys/starknet-snap",
            request: {
              method: "starkNet_switchNetwork",
              params: { chainId: chainId }
            }
          }
        });
        console.log(`Switched to Starknet network with chainId: ${chainId}`);
      } catch (error) {
        console.error("Error switching Starknet network:", error);
        throw error;
      }
    } else {
      console.error("MetaMask not detected or Snaps not supported");
      throw new Error("MetaMask not detected or Snaps not supported");
    }
  };
```

Author: @stanleyyconsensys are we using `window.ethereum` or EIP-6963 ? I don't think we should mix the methods... unless it is absolutely necessary. I think you mentioned that implementing EIP-6963 is a future task and is not fully implemented. Maybe we should wait, instead of partially introducing EIP-6963 for this. WDYT?

---
description: Manage Starknet networks in MetaMask.
sidebar_position: 3
---

# Manage Starknet networks

>  

Reviewer 2: Add brief intro sentence.

## Switch between networks

> Allow users to switch between different Starknet networks, such as mainnet and testnet. This function enables switching by setting the 
> appropriate chain Id. Starknet network has two official networks: 

Reviewer 2: ```suggestion
You can allow users to switch between different Starknet networks using the `wallet_switchStarknetChain` method and setting the 
appropriate chain ID. Starknet has two official networks: 
```

Where is `wallet_switchStarknetChain` documented? is this specific to `get-starknet`?

- Ssepolia testnet chain ID -  `0x534e5f5345504f4c4941`
- Mainnet chain ID -  `0x534e5f4d41494e`

> ```javascript

Reviewer 2: Update code samples to use double quotes and 2-space indents.

const switchChain = async (chainId: string) => {
    try {
        await wallet?.request({
            type: 'wallet_switchStarknetChain',
            params: { chainId: chainId },
        });
        console.log(`Switched to chainId: ${chainId}`);
    } catch (e) {
        console.error("Failed to switch chain:", e);
    }
};
```

> ### Check the current network
> 
> Confirm that the dapp is connected to the correct network:

Reviewer 2: ```suggestion
## Detect the user's network

You can detect the user's currently connect network using the following:
```

Tweak wording to better match https://docs.metamask.io/wallet/how-to/manage-networks/detect-network/

Also maybe this should go before "switch network"?

---
description: Manage Starknet networks in MetaMask.
sidebar_position: 3
---

# Manage Starknet networks
 
## Switch between networks

> Allow users to switch between different Starknet networks, such as mainnet and testnet. This function enables switching by setting the appropriate network RPC URL.

Reviewer 3: ```suggestion
Allow users to switch between different Starknet networks, such as mainnet and testnet. This function enables switching by setting the appropriate chain Id. Starknet network has two official networks: 

- the sepolia testnet, chain id `0x534e5f5345504f4c4941`
- the mainnet, chain id `0x534e5f4d41494e`
```

```javascript

> const switchNetwork = async (network) => {
>   try {
>     const starknet = getStarknet();
> 
>     // Example network URLs
>     const networks = {
>       mainnet: 'https://mainnet.starknet.io',
>       testnet: 'https://sepolia.starknet.io'
>     };
> 
>     // Set the provider based on the selected network
>     starknet.setProvider({ rpcUrl: networks[network] });
> 
>     console.log(`Switched to ${network}:`, starknet.provider);
>   } catch (error) {
>     console.error('Error switching network:', error);
>   }
> };

Reviewer 3: In case we are in the same set up than the tutorial at the beginning and we have access to the wallet object we stored in state. Switching chain is as simple as : 

```suggestion
const switchChain = async (chainId: string) => {
    try {
        await wallet?.request({
            type: 'wallet_switchStarknetChain',
            params: { chainId: chainId },
        });
        console.log(`Switched to chainId: ${chainId}`);
    } catch (e) {
        console.error("Failed to switch chain:", e);
    }
};
```

```

### Check the current network

Confirm that the dapp is connected to the correct network:

```javascript

> const checkCurrentNetwork = () => {
>   const starknet = getStarknet();
>   const currentNetwork = starknet.provider.rpcUrl;
>   console.log('Currently connected to:', currentNetwork);
>   return currentNetwork;
> };

Reviewer 3: ```suggestion
const checkCurrentNetwork = () => {
  return wallet?.chainId;
};
```
